vyper/parser/keccak256_helper.py

```diff
@@ -0,0 +1,69 @@
+
+from vyper.parser.lll_node import (
+    LLLnode,
+)
+from vyper.parser.parser_utils import (
+    getpos,
+    make_byte_array_copier,
+)
+from vyper.types import (
+    BaseType,
+    is_base_type,
+)
+from vyper.utils import (
+    MemoryPositions,
+    bytes_to_int,
+    keccak256,
+)
+
+
+def keccak256_helper(expr, args, kwargs, context):
+    sub = args[0]
+    # Can hash literals
+    if isinstance(sub, bytes):
+        return LLLnode.from_list(
+            bytes_to_int(keccak256(sub)),
+            typ=BaseType('bytes32'),
+            pos=getpos(expr)
+        )
+    # Can hash bytes32 objects
+    if is_base_type(sub.typ, 'bytes32'):
+        return LLLnode.from_list(
+            [
+                'seq',
+                ['mstore', MemoryPositions.FREE_VAR_SPACE, sub],
+                ['sha3', MemoryPositions.FREE_VAR_SPACE, 32]
+            ],
+            typ=BaseType('bytes32'),
+            pos=getpos(expr),
+        )
+    # Copy the data to an in-memory array
+    if sub.location == "memory":
+        # If we are hashing a value in memory, no need to copy it, just hash in-place
+        return LLLnode.from_list(
+            ['with', '_sub', sub, ['sha3', ['add', '_sub', 32], ['mload', '_sub']]],
+            typ=BaseType('bytes32'),
+            pos=getpos(expr),
+        )
+    elif sub.location == "storage":
```

Comment:
@jakerockland I just read this now, after having merged it. Will this not only hash the first 32bytes?

Reply:
@jacqueswww hrm yeah this is an oversight on my part too 😓 

Given that this is the same code generation that is (and was) being used for the built-in `keccak256` function as well, I think the issue here is a bit broader in scope than the string comparison stuff, in that for any hashes of _any_ string literal/bytes array longer than 32 bytes, we are only hashing the first 32 bytes.

Given our definition of the built-in supporting string literals + byte arrays:

```python
def sha3(a) -> b:
  """
  :param a: value to hash
  :type a: either str_literal, bytes, bytes32

  :output b: bytes32
  """
```

I think we may want to open another ticket specifying that we need to fix the `keccak256` behavior all-around to hash _all_ bytes + need to add tests around this both in the context of string comparison but also just using the built-in.

I will have some time this evening that I can get started with writing tests around this to verify the issue + can open up a ticket from there.

Thoughts? 😅

Reply:
Hmm just check again, pretty sure keccak256/sha3 has tests for longer than 32 byte strings.

Reply:
Hrm, yeah you're right we do have a couple of tests for that in `tests/parser/functions/test_keccak256.py`.

I need to do a bit more reading on the LLL here, but it does seem to me that if things currently work properly for hashing byte arrays longer than 32, that string comparison will work as well, as they use the same helper method? 

Likewise, if there was an issue with the string comparison for strings/byte arrays longer than 32, it would make me think we need to add some more tests with the `keccak256` built-in.

Does that make sense? @jacqueswww

Reply:
Either way, it probably would be good to add more nuanced tests to both `functions/test_keccak256.py` and `types/test_string.py`, which I'm happy to do this evening @jacqueswww — from there it will be a lot more clear what exactly is wrong 😅

Reply:
Yeah, clearly morw tricky than one would initially think! :)